### ISSUE TITLE
Hyperparameter-Free Out-of-Distribution Detection Using Softmax of Scaled Cosine Similarity

### ISSUE THREAD

**Issue by jo-kwsm** [jo-kwsm/paper_summary#16]
# INFO
## author
Engkarat Techapanurak<sup>1</sup>, Masanori Suganuma<sup>2,1</sup>, Takayuki Okatani<sup>1,2</sup>

## affiliation
<sup>1</sup>Tohoku University
<sup>2</sup>RIKEN Center 

## conference or year
2019

## link
[arXiv](https://arxiv.org/abs/1905.10628)
[実装](https://paperswithcode.com/paper/hyperparameter-free-out-of-distribution)

# 概要
ood 検出では何かしらのハイパーパラメータを調節する手法が多い。しかしながら、このハイパーパラメータはood によって調整が必要な上に精度が大きく依存してしまう。ood の性質を考えるとこのようなハイパーパラメータが存在するのは不自然である。そこで、本論文ではハイパーパラメータを用いないood 検出手法を提案した。また、他のood 検出では逆伝播の計算が必要なことに対して順伝播の計算のみでよく、スコアに対するノイズを計算して入力に混ぜる必要がないので計算速度が早い。ハイパーパラメータを用いず、速度が早いにもかかわらず、他の手法に十分匹敵する精度を達成した。以下の図は本手法によるid とood の分離。
![Screenshot from 2020-12-28 08-55-55](https://user-images.githubusercontent.com/54238261/103181977-99835400-48ea-11eb-9b3a-f3a2e46c6b63.png)

# 提案手法
以下にモデルの概略図を示す。特徴量抽出後の出力部分に新たな構造を加えることでood 検出を目指す。
![Screenshot from 2020-12-28 09-03-58](https://user-images.githubusercontent.com/54238261/103182078-b10f0c80-48eb-11eb-9b12-447f7ba2e1ad.png)

loss を通常のsoftmaxから、指数部をコサイン類似度に変更する。コサイン類似度計算時に比較対象となる重みはクラスごとにを用意し学習する。
![Screenshot from 2020-12-28 09-05-54](https://user-images.githubusercontent.com/54238261/103182140-6346d400-48ec-11eb-984b-5a51206700d1.png)

また、係数s は以下のように計算し、重みとバイアスは学習対象である。
![Screenshot from 2020-12-28 09-08-27](https://user-images.githubusercontent.com/54238261/103182137-5f1ab680-48ec-11eb-8eb1-76e68299feae.png)

ood 検出は、最大値を示すクラスのコサイン類似度に対してしきい値を設ける。

# 検証
通常の分類精度は落ちる。
![Screenshot from 2020-12-28 09-12-07](https://user-images.githubusercontent.com/54238261/103182182-d18b9680-48ec-11eb-8b61-8815048c083e.png)

ood 検出精度は多くの実験設定でベースラインを超えている。
![Screenshot from 2020-12-28 09-13-34](https://user-images.githubusercontent.com/54238261/103182208-0c8dca00-48ed-11eb-9056-86509624d66a.png)
![Screenshot from 2020-12-28 09-19-11](https://user-images.githubusercontent.com/54238261/103182311-c5ec9f80-48ed-11eb-9f03-7e01f49d0a44.png)

本手法(破線)はハイパーパラメータを用いていないため精度は一定。ベースラインであるmahalanobis 手法はハイパーパラメータ調整時に用いるood dataset によって精度が大きく変わる。
![Screenshot from 2020-12-28 09-15-34](https://user-images.githubusercontent.com/54238261/103182255-7d34e680-48ed-11eb-9b61-9678f750e91f.png)

本手法はid に対して特徴量抽出しづらくなっているが、od に対して頑健になっていることがわかる。
![Screenshot from 2020-12-28 09-21-43](https://user-images.githubusercontent.com/54238261/103182362-21b72880-48ee-11eb-98a5-f14788e1f01b.png)

計算速度が改善している。
![Screenshot from 2020-12-28 09-23-50](https://user-images.githubusercontent.com/54238261/103182420-6fcc2c00-48ee-11eb-8596-8e3ed6038a8e.png)

# 新規性
- ハイパーパラメータを用いない、つまりood データセットを用いずにベースラインに匹敵するood 検出精度。
- 入力に対してスコアに基づくノイズを混ぜないため計算が順伝播だけでよく早い

# 議論，展望

# Comment
- 出力部を変えるだけなので実装が簡単そう
- id への過剰なフィッティングを防ぐことでood に対して頑健になっているのだろうか。id の特徴量をうまく取れている方がood を検知しやすい感覚があったので不思議。
- コサイン類似度をうまく学習する、もしくはうまく利用する工夫があればクラス分類精度かood 検出精度を向上させられそう。
- single fc と w/o WD はよくわからなかった。

# date
Dec 28, 2020
